cerk_port_mqtt_mosquitto/src/port_mqtt.rs

```diff
@@ -0,0 +1,227 @@
+use anyhow::Result;
+use cerk::kernel::{
+    BrokerEvent, CloudEventRoutingArgs, Config, ConfigHelpers, DeliveryGuarantee,
+    IncomingCloudEvent, OutgoingCloudEvent, OutgoingCloudEventProcessed, ProcessingResult,
+};
+use cerk::runtime::channel::{BoxedReceiver, BoxedSender};
+use cerk::runtime::{InternalServerFn, InternalServerFnRefStatic, InternalServerId};
+use cloudevents::{AttributesReader, Event};
+use mosquitto_client::Mosquitto;
+use serde_json;
+use std::collections::HashMap;
+use std::sync::mpsc::{channel, Sender};
+use std::sync::{Arc, Mutex};
+use std::thread;
+use url::Url;
+
+struct Data {
+    unacked: HashMap<i32, String>,
+}
+
+#[derive(Clone)]
+struct Connection {
+    client: Mosquitto,
+    send_topic: Option<String>,
+    send_qos: u8,
+    subscribe_topic: Option<String>,
+    subscribe_qos: u8,
+}
+
+type ArcData = Arc<Mutex<Data>>;
+
+fn build_connection(id: &InternalServerId, config: Config) -> Result<Connection> {
+    let host = config.get_op_val_string("host")?.unwrap();
+    let send_topic = config.get_op_val_string("send_topic")?;
+    let send_qos = config.get_op_val_u8("send_qos")?.unwrap_or(0);
+    let subscribe_topic = config.get_op_val_string("subscribe_topic")?;
+    let subscribe_qos = config.get_op_val_u8("subscribe_qos")?.unwrap_or(0);
```

Comment:
why do we only allow 1 subscribe topic? why is this not an array of objects?

Reply:
it makes the api simpler:
- no two arrays that configure the same thing (topic+qos combo)
- consistent API for `subscribe_x` and `send_x`